### ISSUE TITLE
Feedback regarding Docker

### ISSUE THREAD

**Issue by henfri** [bpereto/borg-hive#32]
Hello,

thanks for providing the docker-compose.yml.

I would consider it useful - especially for users wanting to test, rather than developers - to provide a docker-compose.yml that uses the already built container(s) from the docker hub:
- https://hub.docker.com/r/bpereto/borg-hive
- some borg container available from docker hub
or alternatively (because most will have borg installed already)
the borg installation of the host

Ideal would be, if one would only have to run docker-compose up after adding the /path/to/repo to the .env.

Greetings,
Hendrik

**Comment by bpereto** [comment]
I have the same opinion that `docker-compose.yaml` should use the images from docker hub as it is intended for deployment.
As you may have seen, the docker images on docker hub are brand new and docker-compose.yml is not migrated to those.

If you look at the docker-compose file, the `REPO_PATH` variable  (which is what you mean by `/pat/to/repo`, right?) reflects the path internally in the containers, not where the repo-data is stored.    
The repositories are stored in a docker volume, as defined in the docker-compose.yaml (https://github.com/bpereto/borg-hive/blob/master/docker-compose.yml#L94) which defaults to a local docker volume if not further configured.

**Comment by bpereto** [comment]
> some borg container available from docker hub
or alternatively (because most will have borg installed already)
the borg installation of the host

Because of the close integration of the ssh-daemom, PAM and ldap to query the repository users from borg-hive, there is some customization needed.
Probably, we can use some already existing and maintaned borg images as base for the customization of borg-hive.
Do you have some good recommendations for borg docker images?   
As you may have seen, it would be nice to have multilpe borg versions installed, so that the user in borg-hive can choose per repository, which borg version should be used. #28

**Comment by henfri** [comment]
Hello,

thanks for your reply.

> The repositories are stored in a docker volume, as defined in the docker-compose.yaml (https://github.com/bpereto/borg-hive/blob/master/docker-compose.yml#L94) which defaults to a local docker volume if not further configured.

Ok. Does that work also with remote repositories (via ssh)?

> Do you have some good recommendations for borg docker images?

No, I am using the pre-compiled/statically linked binaries. 

> would be nice to have multilpe borg versions installed, so that the user in borg-hive can choose per repository

Is there really a use-case for that? I have not yet seen a reason not to use the latest borg version.

Regards,
Hendrik

**Comment by henfri** [comment]
Hello again,

it's runnning now. What's the default user/pass?

Regards,
Hendrik

**Comment by bpereto** [comment]
> Ok. Does that work also with remote repositories (via ssh)?

borg-hive is the remote repository. you can access the repository created in borghive with borgbackup. the information is given in borg-hive UI. normally remote borg serves on port 2222 of your machine running docker-compose.    
you can access it with the correct borg syntax: example: `ssh://03m5p6sg@my-borghive-host:test3:2222`

if you let borg-hive run on port 22, the syntax by borgbackup is more simple:    
`borg init --encryption=repokey-blake2 03m5p6sg@my-borghive-host:test3 `

> Is there really a use-case for that? I have not yet seen a reason not to use the latest borg version.

some distribution doesn't have always the latest borg versoin (1.x), which can have some major changes. if borgbackup is always backwards compatible, then the different borg versions are not needed in this use-case.

> it's runnning now. What's the default user/pass?
as described in the documentation, you have to create one manually, there is no default user/pass. this could be automated on first start by the init script.

```
# change into app container
docker exec -it borg-hive_app_1 /bin/bash

# create superuser
./manage.py createsuperuser
```

please be aware that this project is in alpha state. this should outline, that there are features, documentation and production-readiness is missing.

**Comment by henfri** [comment]
Hello,

of course port 22 is already blocked by ssh on the host, so that I need to use the other syntax. Here, the instructions in the web-interface (thanks for the instruction to create the superuser) could be improved

`borg init --encryption=repokey-blake2 yxyaqe8v@localhost:test`
does not work of course.
Instead, this should:
`borg init --encryption=repokey-blake2 ssh://yxyaqe8v@172.22.0.7:2222/test`
But it does not:
`Remote: ssh: connect to host 172.22.0.7 port 2222: Connection refused`

likewise, if I use the IP of the docker host, as you suggested.
`borg init: error: argument REPOSITORY: Invalid location format: "ssh://03m5p6sg@192.168.177.3:test3:2222"
`
Are you sure, it is using port 2222 here? I think I am using the local hosts borg here. 
I took my information from here 
https://github.com/borgbackup/borg/issues/851

I am not convinced that it is a good option to put borg into a container, as this implies port 22 cannot be used. This will cause many questions by users.
Is that not avoidable?

Regards,
Hendrik

**Comment by bpereto** [comment]
>  borg init --encryption=repokey-blake2 ssh://yxyaqe8v@172.22.0.7:2222/test

this looks like the docker container ip, which is not accessible outside of the docker container.

>  borg init: error: argument REPOSITORY: Invalid location format: "ssh://03m5p6sg@192.168.177.3:test3:2222" 

seems to be the wrong format. should this be `ssh://03m5p6sg@192.168.177.3:2222/test' ?

> Are you sure, it is using port 2222 here?

try to reach the port on the shell of the docker host is running (not inside the containers, nor docker, with plain old telnet:
`telnet localhost 2222`

> I am not convinced that it is a good option to put borg into a container, as this implies port 22 cannot be used. This will cause many questions by users.
Is that not avoidable?

yes, you can configure an additional service ip on the docker host (ex. eth0:1) and map this ip in docker-compose to the borg container. Example:   
```
192.168.177.3(eth0 docker host) -> :22 -> sshd from the system
192.168.177.4(eth0:1 docker host) -> :22 -> borg container :22
```
then you can use the standard format of borg.

**Comment by henfri** [comment]
Hello,

```
root@homeserver:/home/henfri# borg init --encryption=repokey-blake2 ssh://03m5p6sg@192.168.177.3:2222/test
Remote: Connection reset by 192.168.177.3 port 2222
Connection closed by remote host. Is borg working on the server?
root@homeserver:/home/henfri# telnet 192.168.177.3 2222
Trying 192.168.177.3...
Connected to 192.168.177.3.
Escape character is '^]'.
SSH-2.0-OpenSSH_7.9p1 Debian-10+deb10u2
exit
Protocol mismatch.
Connection closed by foreign host.
root@homeserver:/home/henfri# netstat -tulpn | grep 2222
tcp6       0      0 :::2222                 :::*                    LISTEN      1382101/docker-prox
```

> yes, you can configure an additional service ip on the docker host (ex. eth0:1) and map this ip in docker-compose to the borg container.

Ah, yes, via MACVLAN, right? I think that that should be the default in the docker-compose.yml.

Regards,
Hendrik

**Comment by bpereto** [comment]
Strange, what are the output of the borg container, when you try to access it?

I spun up a new installation (0.1.1) from scratch:
```
docker-compose up -d db
# wait
docker-compose up

~ >>> sudo netstat -tulpn | grep 2222                                          
tcp6       0      0 :::2222                 :::*                    LISTEN      20301/docker-proxy  

~ >>> borg init --encryption=repokey-blake2 ssh://somerandomuser@localhost:2222/test 
Remote: somerandomuser@localhost: Permission denied (publickey).
Connection closed by remote host. Is borg working on the server?

borg_1     | Invalid user somerandomuser from 172.19.0.1 port 49124
```

Do you have iptables or nftables enabled? Which os are you using?

I would test it without. On my development machine I needed to disable firewalld with nftables:
```
systemctl stop firewalld
systemclt restart docker
```
I'm on arch linux with nftables.

> Ah, yes, via MACVLAN, right? I think that that should be the default in the docker-compose.yml.

hm not quite, I meant to map a specific IP in:
https://docs.docker.com/config/containers/container-networking/
```
-p 192.168.1.100:8080:80

Map TCP port 80 in the container to port 8080 on the Docker host for connections to host IP 192.168.1.100
```

**Comment by henfri** [comment]
Hello,

thanks for your reply and sorry for my late reaction.
```
root@homeserver:~/Docker/borg-hive# docker-compose up -d db
Starting borg-hive_db_1 ... done
root@homeserver:~/Docker/borg-hive# borg init --encryption=repokey-blake2 ssh://somerandomuser@localhost:2222/test
Remote: ssh: connect to host localhost port 2222: Connection refused
Connection closed by remote host. Is borg working on the server?
```
I am still running the version from 8 days ago.
Is it really correct to start the db container only? It does not run borg.


Just to be sure:
```
  borg:
    image: bpereto/borg-hive:borg-0.1.1
    depends_on:
      - db
    env_file:
      - .env
    volumes:
      - borg-repos:/repos
      - borg-config:/config
    ports:
      - "${BORG_SSH_PORT}:22"

more .env
DEBUG=0
#
# docker-compose
#REPO_PATH=/srv/DataPool/Backups/borg/Stoffers
BORG_SSH_PORT=2222
```



```
iptables --version
iptables v1.8.2 (nf_tables)

systemctl stop firewalld
Failed to stop firewalld.service: Unit firewalld.service not loaded.
```

I am running debian
```
root@homeserver:~/Docker/borg-hive# systemctl stop firewalld
Failed to stop firewalld.service: Unit firewalld.service not loaded.
root@homeserver:~/Docker/borg-hive# systemctl stop nftables
Failed to stop nftables.service: Unit nftables.service not loaded.
root@homeserver:~/Docker/borg-hive# systemctl stop iptables
Failed to stop iptables.service: Unit iptables.service not loaded.
```

> hm not quite, I meant to map a specific IP in:

Oh, I was not aware, this is possible.
It's a bit unfortunate, that one needs to create the virtual ip first - makes it harder to set up. 

When starting all, I just noticed:

> watcher_1  |     for filename in os.listdir(current_path):
> watcher_1  | PermissionError: [Errno 13] Permission denied: '/repos/data'

That may have been the problem

I now did:

create a new folder, chmod 777ed it.
Then:
ifconfig eth0:1 192.168.177.66 netmask 255.255.255.0 up

and recreated the container.

> root@homeserver:~/Docker/borg-hive# docker ps | grep 66
> e902e9ff9cec        bpereto/borg-hive:borg-0.1.1   "/init.sh"               4 minutes ago       Up 2 minutes        192.168.177.66:2222->22/tcp   borg-hive_borg_1
> root@homeserver:~/Docker/borg-hive# borg init --encryption=repokey-blake2 ssh://somerandomuser@192.168.177.66:2222/test
> Remote: Connection closed by 192.168.177.66 port 2222
> Connection closed by remote host. Is borg working on the server?

Greetings,
Hendrik

**Comment by bpereto** [comment]
what does docker-compose logs say?

just a complete example done by me with the freshest setup possible:

**remove all data**
```
docker-compose down
sudo docker volume ls | grep 'borg-hive' | xargs  sudo docker volume rm
sudo systemctl stop firewalld
sudo systemctl restart docker
```
** checkout master** (has improvements)
```
# reset all modifications to git
git reset --hard HEAD 
git checkout master
git pull
```

**.env** (set additionally admin pw and debug mode)
```
BORGHIVE_ADMIN_PASSWORD=admin
DEBUG=1

BORG_SSH_PORT=22
MYSQL_DATABASE=borghive
MYSQL_HOST=db
MYSQL_USER=borghive
MYSQL_PASSWORD=borghive
APP_LOG_LEVEL=INFO
DJANGO_ALLOWED_HOSTS=*
BORGHIVE_ADMIN_USER=admin
LDAP_ORGANISATION=borghive
LDAP_DOMAIN=borghive.local
LDAP_ADMIN_PASSWORD=borghive
LDAP_READONLY_USER=true
BORG_LDAP_HOST=ldap://ldap
BORG_LDAP_BASE_DN=dc=borghive,dc=local
BORG_LDAP_READONLY_DN=cn=readonly,dc=borghive,dc=local
BORG_LDAP_READONLY_USER_PASSWORD=readonly
```

**start**
```
docker-compose up
```

* wait for
```
app_1      | *** Operational MODE: single process ***
app_1      | WSGI app 0 (mountpoint='') ready in 1 seconds on interpreter 0x55e6d7019210 pid: 44 (default app)
app_1      | *** uWSGI is running in multiple interpreter mode ***
app_1      | spawned uWSGI worker 1 (and the only) (pid: 44, cores: 1)
```
* init db, create superuser etc. should be done automatically (exceptions can occur on waiting for DB -> its ugly, but not problematic)

**do borg init```
```
~ >>> borg init --encryption=repokey-blake2 ssh://somerandomuser@localhost:2222/test
The authenticity of host '[localhost]:2222 ([::1]:2222)' can't be established.
ECDSA key fingerprint is SHA256:rtRmC9Jm5RpICVmE/G2B+92Fiu3Kx/7OFQNCOu1fnU0.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Remote: Warning: Permanently added '[localhost]:2222' (ECDSA) to the list of known hosts.
Remote: somerandomuser@localhost: Permission denied (publickey).
Connection closed by remote host. Is borg working on the server?
```

**check borg logs**
```
docker-compose logs -f borg
Attaching to borg-hive_borg_1
borg_1     | ssh-keygen: generating new host keys: DSA 
borg_1     | Copy /etc/ssh/ssh_host_rsa_key to /config/ssh_host_rsa_key
borg_1     | Copy /etc/ssh/ssh_host_ecdsa_key to /config/ssh_host_ecdsa_key
borg_1     | Copy /etc/ssh/ssh_host_ed25519_key to /config/ssh_host_ed25519_key
borg_1     | Server listening on :: port 22.
borg_1     | Server listening on 0.0.0.0 port 22.
borg_1     | Invalid user somerandomuser from 172.18.0.1 port 50706
borg_1     | Connection closed by invalid user somerandomuser 172.18.0.1 port 50706 [preauth]
```

So, with a complete fresh setup, the connection is working. I ran out of ideas why your setup is not working. Probably, some hints are in `docker-compose logs -f borg`

**Comment by henfri** [comment]
Hello,

I generally get the same output:
```
root@homeserver:~/Docker/borg-hive# docker-compose down
Removing borg-hive_db_1      ... done
Removing borg-hive_worker_1  ... done
Removing borg-hive_redis_1   ... done
Removing borg-hive_ldap_1    ... done
Removing borg-hive_borg_1    ... done
Removing borg-hive_watcher_1 ... done
Removing borg-hive_app_1     ... done
Removing network borg-hive_default
root@homeserver:~/Docker/borg-hive# sudo docker volume ls | grep 'borg-hive' | xargs  sudo docker volume rm
borg-hive_borg-config
borg-hive_borg-repos
borg-hive_ldap-config
borg-hive_ldap-data
borg-hive_mariadb-data
borg-hive_static
Error: No such volume: local
Error: No such volume: local
Error: No such volume: local
Error: No such volume: local
Error: No such volume: local
Error: No such volume: local
root@homeserver:~/Docker/borg-hive# sudo systemctl stop firewalld
Failed to stop firewalld.service: Unit firewalld.service not loaded.
root@homeserver:~/Docker/borg-hive# sudo systemctl restart docker




had to change ssh port to 2222


root@homeserver:/home/henfri# borg init --encryption=repokey-blake2 ssh://somerandomuser@localhost:2222/test
The authenticity of host '[localhost]:2222 ([127.0.0.1]:2222)' can't be established.
ECDSA key fingerprint is SHA256:rtRmC9Jm5RpICVmE/G2B+92Fiu3Kx/7OFQNCOu1fnU0.
Are you sure you want to continue connecting (yes/no)? yes
Remote: Warning: Permanently added '[localhost]:2222' (ECDSA) to the list of known hosts.
Remote: somerandomuser@localhost: Permission denied (publickey).
Connection closed by remote host. Is borg working on the server?
root@homeserver:/home/henfri#


borg_1     | ssh-keygen: generating new host keys: DSA
borg_1     | Copy /etc/ssh/ssh_host_rsa_key to /config/ssh_host_rsa_key
borg_1     | Copy /etc/ssh/ssh_host_ecdsa_key to /config/ssh_host_ecdsa_key
borg_1     | Copy /etc/ssh/ssh_host_ed25519_key to /config/ssh_host_ed25519_key
borg_1     | Server listening on :: port 22.
borg_1     | Server listening on 0.0.0.0 port 22.
borg_1     | Invalid user somerandomuser from 172.22.0.1 port 56478
borg_1     | Connection closed by invalid user somerandomuser 172.22.0.1 port 56478 [preauth]

```

I did everything as you discribed. But actually, it would have been sufficient to download the docker-compose.yml rather than checking out from git, right?

In the end, we both got:
```
Remote: somerandomuser@localhost: Permission denied (publickey).
Connection closed by remote host. Is borg working on the server?

```

Why is that good?

Regards,
Hendrik

**Comment by bpereto** [comment]
because previous you got `Connection refused`, which seems to be a connection problem.

Now you can go and add an ssh key and create a repository as described in the docs. this gives you a connection string for borgbackup, which should finally work...

**Comment by henfri** [comment]
Ah, ok.
I tried that. Do you see a problem with this key?
`ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQC [xxxx] 18giVrhgrXdbgnRw== root@homeserver.fritz.box
`

It complains that the format is not ssh-rsa AAAAA ... comment formated.
The logs say:
```
app_1      | Unable to decode the key
app_1      | Traceback (most recent call last):
app_1      |   File "/usr/local/lib/python3.8/site-packages/sshpubkeys/keys.py", line 242, in decode_key
app_1      |     decoded_key = base64.b64decode(pubkey_content.encode("ascii"))
app_1      |   File "/usr/local/lib/python3.8/base64.py", line 87, in b64decode
app_1      |     return binascii.a2b_base64(s)
app_1      | binascii.Error: Incorrect padding
app_1      |
app_1      | During handling of the above exception, another exception occurred:
app_1      |
app_1      | Traceback (most recent call last):
app_1      |   File "/app/borghive/lib/validators.py", line 15, in ssh_public_key_validator
app_1      |     key.parse()
app_1      |   File "/usr/local/lib/python3.8/site-packages/sshpubkeys/keys.py", line 424, in parse
app_1      |     self._decoded_key = self.decode_key(pubkey_content)
app_1      |   File "/usr/local/lib/python3.8/site-packages/sshpubkeys/keys.py", line 244, in decode_key
app_1      |     raise MalformedDataError("Unable to decode the key")
app_1      | sshpubkeys.exceptions.MalformedDataError: Unable to decode the key

```

I tried both:
```
ssh-keygen -t rsa -b 4096
ssh-keygen -t rsa -b 2048

```

This had worked previously for me (but with a key I created with putty).

Another feedback regarding the Import:
It would be good if a repository could be Imported without moving it (mapping the data into the container?). My repository is large and I would have to buy more storage to be able to copy it.

Regards,
Hendrik

**Comment by bpereto** [comment]
replace `root@homeserver.fritz.box` with `comment`. problably some parsing bug?

**Comment by henfri** [comment]
No, unfortunately, that did not help.
Is it normal, that "group" is empty and not editable?

Regards,
Hendrik

**Comment by bpereto** [comment]
sorry, missed your last comment.
Group selection is empty if no group is available to select.

**Comment by bpereto** [comment]
closing due to inactivity.
